오픈마켓 [Step 2] shapiro, 승기

shapiro711 opened the pull request:
@jungseungyeo 
안녕하세요 린생! 많이 부족하지만 리뷰를 한번 받고 다시 진행하는것이 좋을것 같아 PR을 보내요!!!
잘부탁드립니다 ^_^

## 구현한 부분
---
1. Collection View, CustomCell UI 구현
2. 서버에서 list 불러오는 부분과 cell을 그려주는 부분 비동기 처리 구현
3. 각 UI layout 적용

## 궁금한 부분
---
1. 현재 스크롤을 계속 내리면 productList에 상품들의 데이터가 계속 쌓이고 있습니다. 유튜브 같은 앱도 새로 위로 올렸을때 새롭게 로딩을 하지 않는것 같아 이전에 로딩했던 page의 정보도 현재 다 가지고 있는데요... 이렇게 구현하는 것이 맞는지 궁금합니다.
2. collectionView의 flowlayout과 cell안에서 layout을 잡아 ui를 구현했습니다. 저희가 layout을 잘 잡은 것인지 감이 안와서 린생의 생각이 궁금합니다!!!
3. 현재 비동기처리 부분에서 성능면으로 손해를 보고 있는 부분이 없는지 궁금합니다.

shapiro711 committed:
feat: Product 타입 구현

통신할 data의 형식에 맞도록 Product 타입의 프로퍼티 및 CodingKey 구현

ohsg0226 committed:
feat: Page 타입 구현

- 통신할 데이터 형식에 맞게 Page 타입의 프로퍼티 및 CodingKey구현

shapiro711 committed:
feat: 파일 정리, test 파일 추가

-MVC 모델에 맞도록 파일 정리
-JSON파일 에셋에 추가
-Test파일 추가

ohsg0226 committed:
feat: Test case 추가

shapiro711 committed:
feat: ThumbnailProduct타입 구현, Test case 추가

- JSON 데이터에 맞도록 ThumbnailProduct 구현
- Items가 제대로 decoding되는지 확인하는 테스트 케이스 추가
- asset name 오타 수정

ohsg0226 committed:
feat: APIModel 생성

-PostItem: POST할 때 사용될 모델
-PatchItem: PATCH할 때 사용될 모델
-DeleteItem: DELETE할 때 사용될 모델

shapiro711 committed:
feat: ParsingManager타입 구현
-decoding을 위한 decodingData 메서드 구현
-encoding을 위한 encodingModel 메서드 구현
-Product타입 프로퍼티 타입 수정

ohsg0226 committed:
feat: NetworkManager 파일 생성

shapiro711 committed:
feat: URLSessionTests파일 생성, MockURLProtocol타입 생성

- test를 위한 setup 설정
- 가짜 통신을 위한 MockURLProtocol 타입 정의

ohsg0226 committed:
feat: APIModelProtocol, ParsingManagerProtocol생성

- multipartform data생성 및 로직 구현

shapiro711 committed:
feat: NetworkManager request메서드구현,  ParsingManager 기능 추가

- ParsingManager multiaprtform 에 맞춰서 data 생성해주는 로직 추가

shapiro711 committed:
refactor: ParsingManager 책임 NetworkManager로 이전

- MultiPartForm데이터를 만들어주는 책임을 NetworkManager로 이전
- request메서드에서 모델만 받고 안에서 처리해주도록 로직 변경
- imageFile 자료구조 변경 Media 프로토콜과 Photo 타입 생성 및 구현
- Boundary 한번만 생성하도록 변경

ohsg0226 committed:
feat: error case 추가

- JsonError 추가
- NetworkError에 invalidModel case 추가

ohsg0226 committed:
style: coding convention 변경

shapiro711 committed:
feat: NetworkManager request 오버로딩 함수 작성, testcase 작성

- 각 http 메서드 성공 test case 구현

ohsg0226 committed:
feat: response가 400일 때 잘못된 모델을 넣은 testcode작성

shapiro711 committed:
refactor: 불필요 코드 삭제

- 테스트를 위해 넣어놨던 불필요한 코드 삭제

ohsg0226 committed:
refactor: 기존 APIModel 삭제 및 APIable 타입 생성

ohsg0226 committed:
refactor: Post,Patch,Get,Delete Api생성

shapiro711 committed:
refactor: NetworkManager request 로직 수정

-contentType에 따른 분기 처리
-request 메서드 바뀐 api모델에 따르도록 로직 수정

ohsg0226 committed:
test: get, post test case 작성

shapiro711 committed:
test: patch, delete, fail test case 추가

ohsg0226 committed:
feat: mainstoryboard에 collectionview, indicator 추가

shapiro711 committed:
feat: customCell 생성

- customCell 생성 및 레이아웃 작업

ohsg0226 committed:
feat: ProductListCollectionView 구현

-network 통신부분 구현
-collectionview datasource구현
-collectionview delegate 구현

shapiro711 committed:
feat: ProductListCustomCell UI구현

-cell에 들어있는 각 뷰 그리는 부분 추가
-imageloading 비동기 처리 추가
-collectionView에서 셀에 맞게 처리 해주는 부분 추가

shapiro711 committed:
style: 코딩 컨벤션 수정

ohsg0226 committed:
style: 코드 컨벤션수정

shapiro711 committed:
Merge branch '3-shapiro' into step-2

jungseungyeo commented on OpenMarket/OpenMarket/Contorller/ProductListViewController.swift:
```diff
@@ -0,0 +1,109 @@
+        let layout = UICollectionViewFlowLayout()
+        layout.scrollDirection = .vertical
+        
+        guard let collectionView = productListCollectionView else { return }
```
여기서 왜 guard문이 필요할까요?

shapiro711 replied:
위에서 IBoutlet으로 받고 있기 때문에 가드문은 따로 필요 없을 것 같습니다!

jungseungyeo commented on OpenMarket/OpenMarket/Contorller/ProductListViewController.swift:
```diff
@@ -0,0 +1,109 @@
+        collectionView.collectionViewLayout = layout
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        view.addSubview(collectionView)
```
스토리보드에서 이미 적용이 되어 있지 않나요? 여기서 addsubView를 해주는 이유가 무엇을까요?

shapiro711 replied:
생각보니 스토리보드에 적용이 되어있어서 필요 없을것 같습니다.

jungseungyeo commented on OpenMarket/OpenMarket/Contorller/ProductListViewController.swift:
```diff
@@ -0,0 +1,109 @@
+        loadListIndicator.startAnimating()
+        self.currentPage = page + 1
+        let apiModel = GetAPI.lookUpProductList(page: page, contentType: .noBody)
+        networkManager.request(apiModel: apiModel) { [self] result in
```
[self]의 역할은 무엇인가요?

shapiro711 replied:
클로저의 캡처 리스트입니다! 어떤것을 캡처할지 그리고 얼마나 강하게 참조할지 (참조타입일 경우) 명시하는 역할을 하는 것으로 알고 있습니다!

jungseungyeo commented on OpenMarket/OpenMarket/Contorller/ProductListViewController.swift:
```diff
@@ -0,0 +1,109 @@
+                    }
+                }
+            case .failure(let error):
+                print(error.localizedDescription)
```
에러 핸드링은 요구 사항에 없긴 하지만 여기서 구현을 어떻게 해주면 좋을까요?

shapiro711 replied:
사용자에게 로딩이 실패했다는 알림창을 띄워주고 다시 로딩을 시도하는식으로 처리하면 좋을것 같아요!

jungseungyeo commented on OpenMarket/OpenMarket/Contorller/ProductListViewController.swift:
```diff
@@ -0,0 +1,109 @@
+extension ProductListViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView,
+                        layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+        return CGSize(width: collectionView.frame.width/2-10, height: collectionView.frame.height/3)
```
```suggestion
        return CGSize(width: collectionView.frame.width / 2-10, height: collectionView.frame.height / 3)
```

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+    
+    static let identifier = "ProductListCustomCollectionViewCell"
+    
+    static func nib() -> UINib {
```
함수가 좋네요! 더 범용적으로 사용하기 위해서 UICollectionViewCell를 extension으로 구현하는 방법으로 사용하면 더욱 가독성을 높을 수있을거 같아요!

shapiro711 replied:
넵 수정해보겠습니다 :)

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+        guard let text = priceLabel.text else { return }
+        let attributedString = NSMutableAttributedString(string: text)
+        let range = (text as NSString).range(of: text)
+        if product.discountedPrice != nil {
```
cell 안에서 뭔가 분기 처리로 값을 세팅 하는 거 같아요. 계산 로직은 controller의 역할이니 controller에게 맡기면 좋을거같아요!

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+    private func loadThumbnails(product: Product) {
+        if let thumbnail = product.thumbnails.first,
+           let thumbnailURL = URL(string: thumbnail),
+           let data = try? Data(contentsOf: thumbnailURL) {
```
URLSession이 아닌 Data(contentOf: URL)를 사용한 이유가 있나요? 두가지 차이 점이 무엇일까요?

shapiro711 replied:
Data(contentOf: URL) 은 url을 통해서 data로 바꿔주는 함수로 알고 있습니다. 그리고 저희는 loadThumbnails 이 호출되기 전에 이미 통신하여 image를 받아왔기 때문에 URLSession이 아니라 Data(contentOf: URL)를 사용했습니다! 질문 의도를 잘 이해를 못한것 같아요....!

YongHoonJJo replied:
Don't use this synchronous initializer to request network-based URLs. 
For network-based URLs, this method can block the current thread for tens of seconds on a slow network, resulting in a poor user experience, and in iOS, may cause your app to be terminated. Instead, for non-file URLs, consider using the dataTask(with:completionHandler:) method of the URLSession class. See Fetching Website Data into Memory for an example.

<https://developer.apple.com/documentation/foundation/nsdata/1413892-init>

여기 참고해보셔요!

jungseungyeo replied:
죠스트 내용을 보고 적용하면 좋을거같아요! URLSession을 활용하여 적용해주세요

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+            stockLabel.textColor = .orange
+        } else {
+            let enoughCount = 9999
+            let leftover = product.stock > enoughCount ? "재고 많음" : "\(product.stock)"
```
뷰에서 계산 로직이 있는게 맞을까요? 혹시 계산 로직의 기준이 어떻게 잡으면 좋을까요?

shapiro711 replied:
저희는 위의 계산 로직이 셀안의 요소가 그려지기 위한 로직으로 생각해서 셀이 담당해도 괜찮을것 같다는 생각을 했었어요! 컨트롤러가 셀이 만들어지기 위한 데이터만 셀에 넘겨주고 셀이 실제로 그려지는 과정은 셀에서 하는것이 역할 분담이 좀 더 확실하다는 생각을 했습니다.! 데이터를 가지고 단순히 그려지기 위한 분기처리 계산 로직은 뷰에 있어도 괜찮다는 생각이 듭니다. 린생은 어떻게 생각하시나요??

jungseungyeo replied:
view의 역할을 보여주기 위해서만 사용하면 좋을거같아요! 뷰의 역할을 뷰를 만들어지는 값을 계산하면 변경 시점을 찾기가 힘들거같아요! controller에서 이 계산이 된 후 stockLabel.text = 값 이런식으로만 넣어주면 좋을거같아요

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+        return UINib(nibName: identifier, bundle: nil)
+    }
+    
+    override func layoutSubviews() {
```
layoutSubviews에서 처리와 awakefromNib()에서 처리하는 차이가 무엇일까요?

shapiro711 replied:
호출되는 시기의 차이가 있는 것으로 알고 있습니다. awakefromNib()은 인스턴스가 생성된 후에 호출이 되고 layoutSubviews()는 레이아웃이 결정되는 과정 중에 처리를 위해 호출되는 것으로 알고있습니다....! 정확한 차이는 좀 더 공부해봐야 할것 같습니다...!

jungseungyeo replied:
https://medium.com/mj-studio/%EB%B2%88%EC%97%AD-ios-%EB%A0%88%EC%9D%B4%EC%95%84%EC%9B%83%EC%9D%98-%EB%AF%B8%EC%8A%A4%ED%84%B0%EB%A6%AC%EB%A5%BC-%ED%8C%8C%ED%97%A4%EC%B9%98%EB%8B%A4-2cfa99e942f9 
여기 블로그를 찾고하면 좋을거 같아요!

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,100 @@
+        self.layer.borderColor = UIColor.lightGray.cgColor
+    }
+    
+    override func prepareForReuse() {
```
super를 쓰고 안쓰고 차이가 있나요?

shapiro711 replied:
차이는 없지만 공식 문서에 super를 호출하는것을 권장하는 것으로 알고있습니다. 고치도록 할게요!! 
The default implementation of this method does nothing. However, when overriding this method, it is recommended that you call super anyway

jungseungyeo commented:
승기, shapiro step2 고생하셨어요!

step2의 대한 리뷰를 남겼어요! 궁금한 거는 언제든지 물어봐주세요!

1. 현재 스크롤을 계속 내리면 productList에 상품들의 데이터가 계속 쌓이고 있습니다. 유튜브 같은 앱도 새로 위로 올렸을때 새롭게 로딩을 하지 않는것 같아 이전에 로딩했던 page의 정보도 현재 다 가지고 있는데요... 이렇게 구현하는 것이 맞는지 궁금합니다.

> 넵 기존의 데이터를 계속 가지고 있으면서 구현을 하게 되는데요 이런 방식을 페이징처리라고 불려요! 

2. collectionView의 flowlayout과 cell안에서 layout을 잡아 ui를 구현했습니다. 저희가 layout을 잘 잡은 것인지 감이 안와서 린생의 생각이 궁금합니다!!!

> 흠.. 현재 오토레이아웃이 깨지고 있는거 같아요!. 오토레이아웃의 값을 지정하는 방식을 알아보는게 중요한거 같아요. 딱 한가지 방법이 있는게 아니라서 이 방법으로 하면 된다 이런거는 없는거 같아요. 

![image](https://user-images.githubusercontent.com/32052386/130646272-e4ebcf24-5979-4921-b9a1-9b46b3d5510d.png)

3. 현재 비동기처리 부분에서 성능면으로 손해를 보고 있는 부분이 없는지 궁금합니다.

> 코멘트에서 남겨놓았습니다! 코드리뷰에서 이유를 적어 주시면 좋을거같아요!

전체적으로 고민한 흔적이 많이 보여요! 아쉬운 부분도 몇 군대 보이는데요. 코멘트를 보면서 이유를 적어주시고 해당 부분의 이상이 없는지 참고 해보면 좋을거같아요! 또 현재 레이아웃이 잘 적용이 안되고 있어요! 레이아웃도 같이 수정하면 좋을거 같아요!

shapiro711 committed:
refactor: 불필요 코드 삭제, 컨벤션 수정

jungseungyeo commented on OpenMarket/OpenMarket/View/ProductListCustomCollectionViewCell.swift:
```diff
@@ -16,17 +16,14 @@ class ProductListCustomCollectionViewCell: UICollectionViewCell {
     }
     
     override func prepareForReuse() {
+        super.prepareForReuse()
```
👍

jungseungyeo commented:
승기, shapiro 수정내용을 잘 하셨습니다! 레이아웃도 잘 맞는거 같은데 작은 해상도에서는 조금 잘 나오지 앖아요 작은 폰의 대해서 대응도 해주시면 좋을거같아요!. 또 URLSession을 활용해서 이미지 loader를 만들어주시면 좋을거같아요!

ohsg0226 committed:
fix: autolayout 깨지는문제 수정

shapiro711 committed:
fix: imageLoad 로직변경, View 계산로직 controller 이동

- loadThumbnails URLSession으로 처리
- CustomCell에서 하던 분기 처리 ProductListViewController로 이동

shapiro711 commented:
@jungseungyeo 
안녕하세요 린생!! 린생이 말씀해주신 부분을 고쳐서 푸시 했습니다 :)
전체적인 구조가 이상한것 같아 마음에 들지 않네요 ㅠㅠ...
프로젝트 마지막 날 이여서 우선 고친내용을 반영하고 잘 마무리 지어보겠습니다
3주 동안 정말 감사합니다 :)

jungseungyeo commented on OpenMarket/OpenMarket/Contorller/ProductListViewController.swift:
```diff
@@ -84,7 +129,17 @@ extension ProductListViewController: UICollectionViewDataSource {
+        
+        cell.updateLabels(title: titleLabel, price: priceLabel, discountdPrice: discountedPriceLabel, stock: stockLabel)
+        cell.loadThumbnails(product: cellProduct) { (image) in
+            cell.thumbnailImage.image = image
```
여기서 넣어주면 객체지향의 은닉화를 위반하는거 같아요! 다음에 한번 고민해보고 적용하면 좋을거같아요!

jungseungyeo reviewed (approved):
승기, shapiro 3주동안 고생하셨어요! 코드를 보면서 항상 발전하는 모습이 좋아요! 고생하셨습니다.
